### PR TITLE
Add workspace options to question settings page

### PR DESCRIPTION
**Pull request by mylesw27**
This PR intends to add workspace options as editable fields in the question settings page.

### CODE REVIEW

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts [hunk 1]
```diff
@@ -305,6 +305,143 @@ export function InstructorQuestionSettings({
...
+                <div class="mb-3">
+                  <label class="form-label" for="workspace_image">Image</label>
+                  <input
+                    type="text"
+                    class="form-control"
+                    id="workspace_image"
+                    name="workspace_image"
+                    value="${resLocals.question.workspace_image}"
+                    ${canEdit ? '' : 'disabled'}
+                  />
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:330]: Do we need to add client-side validation for `image`, `home`, and `port`? Currently if any of workspace settings are set but any of these three are left blank, sync will fail.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 2]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+            questionInfo.workspaceOptions?.rewriteUrl,
+            validatedReqBody.workspace_rewrite_url === 'on',
+            false,
+          ),
+          gradedFiles: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.gradedFiles,
+            validatedReqBody.workspace_graded_files
+              ?.split(',')
+              .map((s) => s.trim())
+              .filter((s) => s !== '') || [],
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:227]: This is doing the correct thing, but would there be a better way to write this? I just want to split the files/directories if multiple exist, trim them so we are not getting unnecessary spaces and store them in an array. If there are no entries, make sure to store an empty array instead of `[""]`.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 3]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+            0,
+          ),
+          home: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.home,
+            validatedReqBody.workspace_home,
+            '',
+          ),
+          args: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.args,
+            shlex.split(validatedReqBody.workspace_args?.replace(/\r\n/g, ' ') || ''),
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:214]: Is this the correct usage of `shlex`? Borrowed from the sync code but want to double check that this is the best way to split the arguments and store them in an array.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 4]
```diff
@@ -123,40 +125,135 @@ router.post(
         );
       }
 
+      const validatedReqBody = z
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:128]: Should I move this to a schema somewhere else? It is only being used once so thought it would be fine to just declare it here and parse.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 3 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:214]: I'd like to tag in @jonatanschroeder, who added support for the array form of args and generally improved things here. I think there's a broader question of how to handle editing of `args`, specifically because we lose track of whether the original format is an array or a string when the JSON is synced to the database.

Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:214]: This is an interesting discussion. Given that the array representation is only based on fixed values and mainly provides a means to make the escaping cleaner, I'd argue that there is value in storing the arguments in array representation in the JSON file in some cases. These may be something like:
1. Any case where escaping is needed (i.e., whenever the joined version includes quotes or escape characters);
2. Any time there are at least two arguments;
3. Always, even if originally they were just a string;
4. Never (i.e., always store the escaped string version), even if originally they were an array.

I'd argue for option 1, but I'm happy with any of these options.

As for editing in a UI, I believe most users using this UI will likely not encounter complex situations for args in workspaces or external graders, so a simple text input (as you are using) may be enough, with a comment that multiple arguments should be separated by spaces and escaped as necessary using the same format as a typical shell. I'm not convinced a textarea is necessary, I don't see a strong need for line breaks, though if you believe there is a reason to add them go for it.

Finally, answering your proper question: there is no need to convert line breaks to spaces, `shlex.split` already takes care of them appropriately.

File: apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts [hunk 5]
```diff
@@ -70,6 +75,26 @@ onDocumentReady(() => {
...
+      return;
+    }
+    try {
+      JSON.parse((e.target as HTMLInputElement).value);
+      workspaceEnvironmentInput?.setCustomValidity('');
+      return;
+    } catch {
+      workspaceEnvironmentInput?.setCustomValidity('Invalid JSON format');
+    }
+  });
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts]: Good catch on these. I have addressed the following in [46c07fb](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/46c07fbccd98c864a07650d74c9ae39219a4df91):
- I have updated `saveButtonEnabling` to include `reportValidity()` if the element is invalid. This is now actually displaying the custom validity message that is being set for any invalid input. 
- I have also update the validation to include invalidate anything that is not an object or is an array (`(typeof value !== 'object' || Array.isArray(value)`). This will now make sure that we cannot submit quoted strings, arrays, booleans or numbers as described above.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Updated to call this `body` in [46c07fb](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/46c07fbccd98c864a07650d74c9ae39219a4df91).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 3 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: After our discussion on Zoom, I have implemented option 2 here for it's simplicity. If there is a space, we will parse with `shlex` and store it as an array. If there is not a space, we will store it as a string. I have also updated the input description with your comment above to reflect this and now reads:
`Command line arguments to pass to the Docker container. Multiple arguments should be separated by spaces and escaped as necessary using the same format as a typical shell.`
Also, updated this from a `<textarea>` to a regular text `<input />`.

Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Here's a different rule to consider: always perform `shlex.split`; if there are multiple strings in the array, store as an array, otherwise store the single item as a string.

This handles these cases consistently (they're contrived, but still):

```
foo'bar'
foo'bar' --baz
```

With the current rules, they're serialized as this:

```
"foo'bar'"
["foobar", "--baz"]
```

With the new rule, we'd consistently apply the escaping:

```
"foobar"
["foobar", "--baz"]
```

All this said: I wonder if we'd simplify things if we just always store things as an array. Are there any downsides to that?

File: apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts]: > I have updated saveButtonEnabling to include reportValidity() if the element is invalid. This is now actually displaying the custom validity message that is being set for any invalid input.

IMO it feels like it's going beyond the responsibility of that bit of code to be dealing with form validity. I think it's also considered a valuable accessibility feature to validate the form _on_ an attempted submit and not just disable the submit button entirely.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 3 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: I believe this would only really impact anyone that cares about using a string in their JSON file instead of an array. As far as performance, I believe we can safely always store an array. I have updated to this in [184e53d](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/184e53d3f0943723f9aee81bf6fba00cbf3029e2) so we can see what this looks like. It certainly is simpler from a code perspective.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 6]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+          orig_hash: z.string(),
+          qid: z.string(),
+          title: z.string(),
+          topic: z.string().optional(),
+          tags: z.union([z.string(), z.array(z.string())]).optional(),
+          grading_method: z.string().optional(),
+          single_variant: z.string().optional(),
+          show_correct_answer: z.string().optional(),
+          workspace_image: z.string().optional(),
+          workspace_port: z.coerce.number().optional(),
```
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:141]: I think we are going to treat this as an integer below? If so, this can be https://github.com/PrairieLearnInc/PrairieTest/blob/10756bb41894ef4d19ce4865a6e5376790a81148/src/lib/zod-types.ts#L18 (this will need to be lifted over to `packages/zod` in PL)

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 7]
```diff
@@ -123,40 +125,135 @@ router.post(
         );
       }
 
+      const body = z
+        .object({
+          __action: z.string(),
+          __csrf_token: z.string(),
+          orig_hash: z.string(),
+          qid: z.string(),
```
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:133]: We could eliminate the check for the presence of a QID above and shift the format check for the QID to below this body parsing step.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 8]
```diff
@@ -123,40 +125,135 @@ router.post(
         );
       }
 
+      const body = z
```
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:128]: I like defining the body parsing within a particular `__action` section so we know what to expect ✅

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 9]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+            0,
+          ),
+          home: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.home,
+            body.workspace_home,
+            '',
+          ),
+          args: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.args,
+            shlex.split(body.workspace_args || ''),
```
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:214]: Define a zod schema for this?

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 10]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+            questionInfo.workspaceOptions?.rewriteUrl,
+            body.workspace_rewrite_url === 'on',
+            false,
+          ),
+          gradedFiles: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.gradedFiles,
+            body.workspace_graded_files
+              ?.split(',')
+              .map((s) => s.trim())
+              .filter((s) => s !== '') || [],
```
Comment (mwest1066) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:227]: Define a zod schema?

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 4 above)
Comment (mwest1066) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: I left a bunch of comments. In general I think that it's better to define schemas for things like this in `packages/zod` for two reasons:
1. We often end up having to reuse these.
2. By defining an actual schema you can also add some tests to make sure that it's working as expected.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 11]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+          grading_method: z.string().optional(),
+          single_variant: z.string().optional(),
+          show_correct_answer: z.string().optional(),
+          workspace_image: z.string().optional(),
+          workspace_port: z.coerce.number().optional(),
+          workspace_home: z.string().optional(),
+          workspace_args: z.string().optional(),
+          workspace_rewrite_url: z.string().optional(),
+          workspace_graded_files: z.string().optional(),
+          workspace_enable_networking: z.string().optional(),
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Updated to use `BooleanFromCheckboxSchema` for the checkboxes in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 12]
```diff
@@ -123,40 +125,135 @@ router.post(
...
+      const body = z
+        .object({
+          __action: z.string(),
+          __csrf_token: z.string(),
+          orig_hash: z.string(),
+          qid: z.string(),
+          title: z.string(),
+          topic: z.string().optional(),
+          tags: z.union([z.string(), z.array(z.string())]).optional(),
+          grading_method: z.string().optional(),
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Added `EnumGradingMethodSchema` to `db-types` in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153). Updated to use it there and use it here to define `grading_method`.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 6 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Added `IntegerFromStringSchema` to our `zod` package in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153). This can also come through as an empty string if a workspace is not defined. To fix this, I also added `IntegerFromStringOrEmptySchema` which acts as a wrapper to convert an empty string to a `0` then pass the value to `IntegerFromStringOrEmpty`.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 13]
```diff
@@ -123,40 +125,135 @@ router.post(
...
 
+      const body = z
+        .object({
+          __action: z.string(),
+          __csrf_token: z.string(),
+          orig_hash: z.string(),
+          qid: z.string(),
+          title: z.string(),
+          topic: z.string().optional(),
+          tags: z.union([z.string(), z.array(z.string())]).optional(),
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Added `ArrayFromStringOrArraySchema` to our package in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 14]
```diff
@@ -123,40 +125,135 @@ router.post(
         );
       }
 
+      const body = z
+        .object({
+          __action: z.string(),
+          __csrf_token: z.string(),
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Removed these properties in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 9 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: I don't know when we will use this schema again, so I defined this inline in the body schema in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153). Line 137 is now:
```
workspace_args: z.string().transform((s) => shlex.split(s || '')),
```

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 10 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Similar to `workspace_args` I don't know if we need a reusable schema for this. Instead I moved this inline to the body schema as well. Lines 139 - 144 are now:
```
workspace_graded_files: z.string().transform((s) =>
    s
        .split(',')
        .map((s) => s.trim())
        .filter((s) => s !== ''),
    ),
```

File: apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts [hunk 15]
```diff
@@ -70,6 +75,30 @@ onDocumentReady(() => {
...
+    } catch {
+      workspaceEnvironmentInput?.setCustomValidity('Invalid JSON object format');
+    }
+  });
+
   if (!questionSettingsForm || !saveButton) return;
   saveButtonEnabling(questionSettingsForm, saveButton);
+
+  showWorkspaceOptionsButton?.addEventListener('click', (e) => {
+    e.preventDefault();
```
Comment (nwalters512) [inline apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts:100]: Is this needed?

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts [hunk 16]
```diff
@@ -305,6 +305,145 @@ export function InstructorQuestionSettings({
...
+                    id="workspace_home"
+                    name="workspace_home"
+                    value="${resLocals.question.workspace_home}"
+                    ${canEdit ? '' : 'disabled'}
+                  />
+                  <small class="form-text text-muted">
+                    The home directory of the workspace container.
+                  </small>
+                </div>
+                <div class="mb-3">
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:364]: Nit, can we reorder these inputs? I'd like those that are going to be most commonly-used to be first. Image/port/home are already good. Let's follow that with graded files (will be specified for most questions), arguments, env, enable networking, rewrite URL (should almost never be used).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts:330]: That's a good point. I think we should add that. I think we'd have to conditionally add `required` to those inputs when the workspace options are expanded.

File: apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts]: @mylesw27 I wanted to make sure this comment didn't get lost!

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts (hunk 1 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts]: If we add required when the workspace options are expanded, there is no way to remove a configured workspace. Instead I am now conditionally adding `required` to those inputs if any of the workspace inputs contain any values in [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 2 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: I believe this is correct. Added a Zod validation for this in [965d72c](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/965d72c605ecd80ce579faeb619594cfc5d5e153).

File: apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts]: Sorry, I had missed it! In [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac), I removed the form validity pieces from this function and we are now checking form validity on submission rather than disabling the button and evaluating on each change.

File: packages/zod/src/index.ts [hunk 17]
```diff
@@ -102,3 +102,48 @@ export const DateFromISOString = z
...
+      message: 'must be a valid integer',
+    },
+  )
+  .transform((s) => Number.parseInt(s));
+
+/**
+ * A Zod schema that utilizes IntegerFromStringSchema to parse a string as an
+ * integer. However, this wrapper also allows us to accept and empty string as a
+ * valid value. If an empty string is passed, it will be transformed to 0.
+ * This is useful for form inputs where an empty string is a valid.
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply packages/zod/src/index.ts]: Updated this comment in [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac).

File: packages/zod/src/index.ts
[the remark this answers is not in the document]
Comment (mylesw27) [reply packages/zod/src/index.ts:1]: Ran `make changeset` in [baf2801](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/baf2801e69c117ddd66335c99011d2c400f80fcb).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 18]
```diff
@@ -115,126 +121,120 @@ router.post(
...
-          grading_method: z.string().optional(),
-          single_variant: z.string().optional(),
-          show_correct_answer: z.string().optional(),
+          tags: ArrayFromStringOrArraySchema.optional(),
+          grading_method: EnumGradingMethodSchema.optional(),
+          single_variant: BooleanFromCheckboxSchema,
+          show_correct_answer: BooleanFromCheckboxSchema,
           workspace_image: z.string().optional(),
-          workspace_port: z.coerce.number().optional(),
+          workspace_port: IntegerFromStringOrEmptySchema,
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: No, updated this to return `null` if it is an empty string. This is in alignment with the change below that the default value for this field should be `null`.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 19]
```diff
@@ -127,36 +158,102 @@ router.post(
...
+          comment: questionInfo.workspaceOptions?.comment ?? undefined,
+          image: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.image,
+            body.workspace_image,
+            '',
+          ),
+          port: propertyValueWithDefault(
+            questionInfo.workspaceOptions?.port,
+            body.workspace_port,
+            0,
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Updated this to be `null` in [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 20]
```diff
@@ -113,10 +121,33 @@ router.post(
...
+          workspace_port: IntegerFromStringOrEmptySchema,
+          workspace_home: z.string().optional(),
+          workspace_args: z.string().transform((s) => shlex.split(s || '')),
+          workspace_rewrite_url: BooleanFromCheckboxSchema,
+          workspace_graded_files: z.string().transform((s) =>
+            s
+              .split(',')
+              .map((s) => s.trim())
+              .filter((s) => s !== ''),
+          ),
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:147]: Added a comment that we know this does not handle filenames with commas in them in [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac).

File: apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts (hunk 15 above)
Comment (mylesw27) [reply apps/prairielearn/assets/scripts/instructorQuestionSettingsClient.ts]: Removed in [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts (hunk 16 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.ts]: Reordered to this suggested order in [3d576d0](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/3d576d0435873caf3c07dc9b4552ebcc178a41ac).

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 9 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Inline schemas are totally fine if we don't yet have a reason to reuse 👍

File: packages/zod/src/index.ts
Comment (nwalters512) [inline packages/zod/src/index.ts:1]: As discussed on our call, let's add some tests (both positive and negative) for these new schemas.

File: packages/zod/src/index.ts [hunk 21]
```diff
@@ -102,3 +102,61 @@ export const DateFromISOString = z
...
+  .refine(
+    (s) => {
+      const n = Number.parseInt(s);
+      return !Number.isNaN(n) && Number.isSafeInteger(n);
+    },
+    {
+      message: 'must be a valid integer',
+    },
+  )
+  .transform((s) => Number.parseInt(s));
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply packages/zod/src/index.ts:121]: cc @mwest1066 I'm wondering if the above is simple enough to just inline whenever we care about this?

File: packages/zod/src/index.ts [hunk 22]
```diff
@@ -102,3 +102,61 @@ export const DateFromISOString = z
...
+        return true;
+      }
+      const n = Number.parseInt(s);
+      return !Number.isNaN(n) && Number.isSafeInteger(n);
+    },
+    {
+      message: 'must be a valid integer',
+    },
+  )
+  .transform((s) => (s === null ? null : Number.parseInt(s)));
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply packages/zod/src/index.ts]: Updated with Prettier and added tests in [2d6517d](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/2d6517d018158a762a82300eb28a19f7a986aaee).

File: packages/zod/src/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (mylesw27) [reply packages/zod/src/index.ts]: Updated this to `z.coerce.number().int()` in [2d6517d](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/2d6517d018158a762a82300eb28a19f7a986aaee).

File: packages/zod/src/index.ts
Comment (mylesw27) [reply packages/zod/src/index.ts:1]: Added tests for the integer schemas in [2d6517d](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/2d6517d018158a762a82300eb28a19f7a986aaee) and the array schema in [676e5e9](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/676e5e97ee26d42bb25467d43edd808f01057f69).

Comment (nwalters512) [reply packages/zod/src/index.ts:1]: Looks great, thank you!

File: packages/zod/src/index.ts [hunk 23]
```diff
@@ -102,3 +102,38 @@ export const DateFromISOString = z
...
+export const IntegerFromStringSchema = z.coerce.number().int();
+
+/**
+ * A Zod schema that utilizes IntegerFromStringSchema to parse a string as an
+ * integer. However, this wrapper also allows us to accept and empty string as a
+ * valid value. If an empty string is passed, it will be transformed to null.
+ * This is useful for form number inputs that are not required but we do not want to
+ * use an empty string to compute values.
+ */
+
```
Comment (nwalters512) [inline packages/zod/src/index.ts:119]: Tiny nit:
```suggestion
```

File: apps/prairielearn/src/tests/instructorQuestionSettings.test.ts [hunk 24]
```diff
@@ -79,6 +79,13 @@ describe('Editing question settings', () => {
         qid: 'question',
         topic: 'Test2',
         tags: 'test2',
+        grading_method: 'Internal',
+        workspace_image: '',
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/instructorQuestionSettings.test.ts:83]: Aside: I noticed that `workspace_image` was declared with `.optional()` but the others weren't. In theory they could all be optional to keep the tests a little more concise? Writing this out I'm not sure if it's worth it, but leaving this comment anyways for your consideration.

Regardless: I think we should treat the workspace options the same way (either all optional, or all not).

File: packages/zod/src/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (mwest1066) [reply packages/zod/src/index.ts]: I’m good with this

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts [hunk 25]
```diff
@@ -127,36 +161,102 @@ router.post(
...
+      if (
+        body.workspace_image ||
+        body.workspace_port ||
+        body.workspace_home ||
+        body.workspace_args ||
+        body.workspace_rewrite_url ||
+        body.workspace_graded_files ||
+        body.workspace_enable_networking ||
+        body.workspace_environment
+      ) {
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts:200]: I got rid of this condition (that is, just unconditionally computed `workspaceOptions`) and the tests still passed. Is this actually needed? If yes, we should be able to add a test that will fail without this. If not: let's just drop it.

File: apps/prairielearn/src/lib/editors.ts [hunk 26]
```diff
@@ -185,12 +185,19 @@ export function getUniqueNames({
  * than the default value.
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/editors.ts:185]: Nit, clarifying docs here:

```suggestion
 * than the default value.
 *
 * `defaultValue` may be either a value to compare directly with `===`, or a function
 * that accepts a value and returns a boolean to indicate if it should be considered
 * a default value.
```

**Review by nwalters512 — COMMENTED** [review]
A few more notes, sorry about that!

File: packages/zod/src/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply packages/zod/src/index.ts]: 👍 in that case, @mylesw27 want to remove this special schema in favor of the above? You can make the same change in PT when you go to replace its schemas with the new shared ones.

File: apps/prairielearn/src/tests/instructorQuestionSettings.test.ts (hunk 24 above)
Comment (mylesw27) [reply apps/prairielearn/src/tests/instructorQuestionSettings.test.ts]: That makes sense! Updated to make all workspace options optional in [a5f6ce4](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/a5f6ce43bff8656480918452dec50b76d5145d95) and removed the now unnecessary lines from the tests.

File: apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts (hunk 25 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.ts]: Originally, this was intended to keep from needlessly evaluating workspace options if there were not any set but is not necessary. Removed in [a5f6ce4](https://github.com/PrairieLearn/PrairieLearn/pull/11811/commits/a5f6ce43bff8656480918452dec50b76d5145d95).

File: packages/zod/src/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply packages/zod/src/index.ts]: Looks like this has been addressed ✅

File: apps/prairielearn/assets/scripts/lib/saveButtonEnabling.ts [hunk 27]
```diff
@@ -25,7 +25,12 @@ export function saveButtonEnabling(form: HTMLFormElement, saveButton: HTMLButton
   // set valueHasChanged{} to false. Then call checkDifferences() to see if the save
   // button should be enabled.
   form.addEventListener('input', (e) => {
-    if ((e.target as HTMLInputElement).value === (e.target as HTMLInputElement).defaultValue) {
+    if (
+      ((e.target as HTMLInputElement).type === 'checkbox' &&
+        (e.target as HTMLInputElement).checked === (e.target as HTMLInputElement).defaultChecked) ||
+      ((e.target as HTMLInputElement).type !== 'checkbox' &&
+        (e.target as HTMLInputElement).value === (e.target as HTMLInputElement).defaultValue)
+    ) {
```
Comment (mylesw27) [inline apps/prairielearn/assets/scripts/lib/saveButtonEnabling.ts:33]: As pointed out by @nwalters512 on Slack, these inputs were not correctly handling checkboxes. I have updated to evaluate checkboxes based on `checked` rather than `value`. Checkboxes like this always have a `value` of `on` regardless of checked status.

Comment (nwalters512) [reply apps/prairielearn/assets/scripts/lib/saveButtonEnabling.ts]: Looks great! I cleaned things up slightly in 5384d95.

**Review by nwalters512 — APPROVED** [review]
